### PR TITLE
NumPy Copy/View Semantics

### PR DESCRIPTION
**Pull request by alexnick83**
This PR amends slicing of DaCe Arrays in the Python frontend so that it follows NumPy rules. For example, simple slicing should generate DaCe Views.

### CODE REVIEW

File: dace/frontend/python/newast.py [hunk 1]
```diff
@@ -1561,6 +1580,9 @@ def parse_program(self, program: ast.FunctionDef, is_tasklet: bool = False):
             if vname.startswith('__return'):
                 if isinstance(self.sdfg.arrays[arrname], data.View):
                     # In case of a view, make a copy
+                    # NOTE: If we are at the top level SDFG (not always clear),
+                    # and it is a View of an input array, can we return a NumPy
+                    # View directly?
```
Comment (tbennun) [inline dace/frontend/python/newast.py:1585]: probably not, a non-transient array acts as a view

File: dace/frontend/python/replacements.py [hunk 2]
```diff
@@ -4305,19 +4353,21 @@ def reshape(pv: 'ProgramVisitor',
...
+                                    total_size=desc.total_size,
+                                    may_alias=desc.may_alias,
+                                    alignment=desc.alignment,
+                                    find_new_name=True)
 
     # Register view with DaCe program visitor
-    pv.views[newarr] = arr
+    aset = subsets.Range.from_array(desc)
+    vset = subsets.Range.from_array(newdesc)
+    pv.views[newarr] = (arr, Memlet(f'{arr}[{aset}]->{vset}'))
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:4370]: would be cleaner and safer to use `Memlet(data=arr, subset=aset, other_suset=vset)`

File: dace/frontend/python/replacements.py [hunk 3]
```diff
@@ -4410,17 +4463,19 @@ def flat(pv: 'ProgramVisitor', sdfg: SDFG, state: SDFGState, arr: str) -> str:
...
+                                        allow_conflicts=desc.allow_conflicts,
+                                        total_size=totalsize,
+                                        may_alias=desc.may_alias,
+                                        alignment=desc.alignment,
+                                        find_new_name=True)
         # Register view with DaCe program visitor
-        pv.views[newarr] = arr
+        aset = subsets.Range.from_array(desc)
+        vset = subsets.Range.from_array(newdesc)
+        pv.views[newarr] = (arr, Memlet(f'{arr}[{aset}]->{vset}'))
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:4478]: same

File: tests/numpy/copies_and_views_test.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import dace
```
Comment (tbennun) [inline tests/numpy/copies_and_views_test.py:1]: copyright missing

File: tests/numpy/copies_and_views_test.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    is_a_copy.f(ref)
+    assert (np.allclose(val, ref))
+
+
+if __name__ == '__main__':
+    test_set_by_view()
+    test_set_by_view_1()
+    test_set_by_view_2()
+    test_set_by_view_3()
+    test_is_a_copy()
```
Comment (tbennun) [inline tests/numpy/copies_and_views_test.py:82]: newline missing

**Review by tbennun — CHANGES_REQUESTED** [review]
cool PR, minor changes only

File: dace/frontend/python/newast.py [hunk 6]
```diff
@@ -1172,7 +1174,7 @@ def _add_access(
             arr_type = type(parent_array)
         if arr_type == data.Scalar:
             self.sdfg.add_scalar(var_name, dtype)
-        elif arr_type == data.Array:
+        elif arr_type in (data.Array, data.View):
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:1177]: The `_add_access` methods are adding outer-level accesses to nested scopes/SDFGs, i.e., they are creating connectors. Therefore, such arrays will be pointers (and not copies) in C/C++ code anyway. Do you think that there is any reason (potentially for some kind of future use) to make connectors to Views be Views themselves? Will this even work properly at the moment?

File: dace/frontend/python/newast.py (hunk 1 above)
Comment (alexnick83) [reply dace/frontend/python/newast.py:1585]: I am considering two cases here that might be interesting to users (and probably difficult to implement properly, if possible at all):

The first case is returning a View of an input. For example:
```python
def foo(x):
    return x[1:5]
    
A = np.arange(10)
B = foo(A)
assert (B.base is A)
```

The second case is returning a View of transient data:
```python
def foo():
    x = np.arange(10)
    return x[1:5]

B = foo()
assert (B.base is not B)
```

Any work on such cases will have to be done on another PR. However, do you think that there would be any benefit in trying to support such codes or is it better to discourage the users from such coding practices?

File: dace/frontend/python/replacements.py (hunk 2 above)
Comment (alexnick83) [reply dace/frontend/python/replacements.py:4370]: I thought that we actually preferred the string representation but it is easy to swap.

File: dace/frontend/python/newast.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/frontend/python/newast.py:1177]: Sounds good

File: dace/frontend/python/newast.py (hunk 1 above)
Comment (tbennun) [reply dace/frontend/python/newast.py:1585]: No, we do want that behavior possible and we can assume that a copy will be made, as long as it's documented

File: dace/frontend/python/replacements.py (hunk 2 above)
Comment (tbennun) [reply dace/frontend/python/replacements.py:4370]: No, we prefer the raw objects as much as possible (faster performance, less ambiguity). The strings are mostly for convenience.

**Review by tbennun — APPROVED** [review]
LGTM!

File: tests/symbol_dependent_transients_test.py [hunk 7]
```diff
@@ -144,6 +172,27 @@ def test_symbol_dependent_pinned_array():
     assert (np.allclose(B, B_ref))
 
 
+@pytest.mark.skip  # @pytest.mark.gpu
+def test_symbol_dependent_gpu_view():
+    # NOTE: This test cannot produce the correct result since the input
+    # data of the reduction are not contiguous and cub:reduce doesn't support
```
Comment (tbennun) [inline tests/symbol_dependent_transients_test.py:178]: can't we force expansion to be pure here? we do this in some of the library states with a `with` statement

Comment (alexnick83) [reply tests/symbol_dependent_transients_test.py:178]: Maybe the test should just change. It is not that important at this time.

**Pull request merged**